### PR TITLE
add Lfm2_VL

### PR DESCRIPTION
**Pull request by yicycyc**
## PR 新增模型支持：LFM2-VL

本 PR 新增 `LiquidAI/LFM2.5-VL-450M` 支持。复现对齐对象为 HuggingFace Transformers 中的 LFM2-VL 实现及原始 safetensors 权重，验证使用完整的 450M checkpoint。

## 主要改动

- 新增 `paddleformers/transformers/lfm2_vl/`
  - `Lfm2Config`
  - `Lfm2Model`
  - `Lfm2ForCausalLM`
  - `Lfm2VlConfig`
  - `Lfm2VlModel`
  - `Lfm2VlForConditionalGeneration`
  - `Siglip2VisionConfig`
  - `Siglip2VisionModel`
  - `Lfm2VlProcessor`
  - `Lfm2VlImageProcessor`
  - LFM2 full-attention/short-convolution hybrid text backbone
  - SigLIP2 NaFlex 视觉编码器、多模态 projector 和图文特征合并
- 注册 LFM2 和 LFM2-VL 到：
  - `paddleformers.transformers`
  - `AutoConfig`
  - `AutoModelForCausalLM`
  - `AutoModelForConditionalGeneration`
  - `AutoProcessor`
  - `AutoImageProcessor`
- 权重加载：
  - 支持直接加载原始 HuggingFace safetensors 目录
  - 自动转换 text、vision 和 projector 中的 Linear 权重，并校验 missing、unexpected 和 shape mismatch
  - 用户无需预先运行离线权重转换脚本
- 新增 `tests/transformers/lfm2_vl/test_modeling.py`
  - 覆盖多模态 forward/loss/backward、纯文本 forward、AutoModel 加载和图像预处理

## 前向对齐验证

模型：`LiquidAI/LFM2.5-VL-450M`

验证设置：

- Transformers 与 PaddleFormers 加载相同原始权重和相同图文输入
- 单张 NVIDIA A100 40GB
- FP32
- attention backend：两侧均使用 eager
- 输出 logits shape：`[1, 96, 65536]`

结果：

| 对比范围 | max abs diff | mean abs diff | p99 abs diff | cosine | Top-1 一致率 |
| --- | ---: | ---: | ---: | ---: | ---: |
| 完整图文序列 | `2.13e-3` | `5.88e-5` | `5.79e-4` | `0.9999999999` | `100%` |
| 有效文本位置 | `2.67e-4` | `2.23e-5` | `2.22e-4` | `0.999999999998` | `100%` |

前向对齐通过。

## 生成对齐

使用相同权重和输入进行 greedy generation，关闭采样，两侧前 10 个新 token 完全一致。

结论：前 10 个 token 匹配率为 `10/10`，生成对齐通过。

## 训练验证

PaddleFormers 与 ms-swift 使用相同的预处理结果和样本顺序，在完整 `LiquidAI/LFM2.5-VL-450M` 上完成单卡 BF16 full-parameter SFT。

共同设置：

- 单张 NVIDIA A100 40GB
- `max_steps = 300`
- `per_device_train_batch_size = 1`
- `gradient_accumulation_steps = 1`
- `learning_rate = 1e-5`
- `lr_scheduler_type = constant`
- `warmup_steps = 0`
- `weight_decay = 0`
- gradient clipping：global norm `1.0`
- `seed = 42`
- shuffle 关闭
- attention eager
- full-parameter SFT

### 1、纯文本

使用 GSM8K 300 条固定样本做 300-step full-SFT。

| 指标 | ms-swift | PaddleFormers | Paddle - ms-swift |
| --- | ---: | ---: | ---: |
| final loss | `0.99505` | `0.87285` | `-0.12220` |
| 300-step mean loss | `0.74447` | `0.69170` | `-0.05277` |
| last 50 mean loss | `0.74303` | `0.70497` | `-0.03806` |
| loss correlation | — | — | `0.94719` |

gsm_loss_comparison

### 2、多模态

使用 PaddleFormers 多模态 SFT 数据格式准备 300 条固定图文样本，两侧使用完全相同的 token、label、图像张量和样本顺序。

| 指标 | ms-swift | PaddleFormers | Paddle - ms-swift |
| --- | ---: | ---: | ---: |
| final loss | `2.78877` | `2.78642` | `-0.00236` |
| 300-step mean loss | `1.91723` | `1.88369` | `-0.03354` |
| last 50 mean loss | `1.72192` | `1.69846` | `-0.02346` |
| loss correlation | — | — | `0.98683` |

loss_comparison

纯文本和多模态训练均跑满 300 step，两侧原始 loss 收敛趋势一致。

### CODE REVIEW

File: paddleformers/transformers/lfm2_vl/modeling.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, *args, **kwargs):
+        checkpoint_path = (
+            os.path.join(pretrained_model_name_or_path, "model.safetensors")
+            if isinstance(pretrained_model_name_or_path, str)
+            else ""
+        )
+        if not os.path.isfile(checkpoint_path):
+            return super().from_pretrained(pretrained_model_name_or_path, *args, **kwargs)
```
Comment (risemeup1111) [inline paddleformers/transformers/lfm2_vl/modeling.py:375]: P1 这里仅在传入本地目录且其中存在 `model.safetensors` 时执行自定义 HF 权重转换；标准调用 `Lfm2VlForConditionalGeneration.from_pretrained("LiquidAI/LFM2.5-VL-450M")` 会落入基类路径，而基类默认使用 flex checkpoint 并要求 `_gen_aoa_config`，当前类未实现该方法，加载会在构造模型后直接报错。请让 Hub/分片 safetensors 也走同一转换逻辑，或提供完整的 `_gen_aoa_config`/非 flex 加载路径。

File: paddleformers/transformers/lfm2_vl/modeling.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+            raise ValueError("Specify exactly one of input_ids and inputs_embeds")
+        if inputs_embeds is None:
+            inputs_embeds = self.get_input_embeddings()(input_ids)
+        image_features = None
+        if pixel_values is not None:
+            image_features_list = self.get_image_features(
+                pixel_values, spatial_shapes, pixel_attention_mask
+            ).pooler_output
+            image_features = paddle.concat(image_features_list, axis=0).astype(inputs_embeds.dtype)
+            image_mask = input_ids == self.config.image_token_id
```
Comment (risemeup1111) [inline paddleformers/transformers/lfm2_vl/modeling.py:317]: P1 该分支允许 `inputs_embeds` 非空且 `input_ids=None`，但这里仍直接对 `input_ids` 做比较；多模态调用传入 `inputs_embeds + pixel_values` 时会得到 Python `bool`，下一行调用 `.astype` 即报错，无法完成图像特征注入。请在 `input_ids` 缺失时根据图像 token 的 embedding 构造 placeholder mask（或明确拒绝这一受支持的输入组合），并补回归测试。

File: paddleformers/transformers/lfm2_vl/processor.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                if prompt.count(self.image_token) != 1:
+                    raise ValueError("Each LFM2-VL image prompt must contain exactly one <image> token.")
+                replacement = self._tokens_per_image(
+                    image_inputs["image_rows"][index],
+                    image_inputs["image_cols"][index],
+                    image_inputs["image_sizes"][index],
+                    use_image_special_tokens,
+                )
+                texts[index] = prompt.replace(self.image_token, replacement)
+        tokenized = self.tokenizer(texts, return_tensors=return_tensors, **kwargs)
```
Comment (risemeup1111) [inline paddleformers/transformers/lfm2_vl/processor.py:71]: P2 当文本含有 `<image>` 但 `images=None` 时，这里会直接把占位符交给 tokenizer；模型随后走 text-only 分支，用普通 embedding 静默替代图像特征，输出会与用户输入语义不符。请在处理器中像 HF 实现一样检测“有图像 token 但没有图像输入”并抛出明确错误，同时覆盖该场景测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [modeling.py:284](https://github.com/PaddlePaddle/PaddleFormers/pull/4865#discussion_r3755583208) | P1 | ✅ |
| 2 | [modeling.py:246](https://github.com/PaddlePaddle/PaddleFormers/pull/4865#discussion_r3755584905) | P1 | ✅ |
| 3 | [processor.py:69](https://github.com/PaddlePaddle/PaddleFormers/pull/4865#discussion_r3755586147) | P2 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/lfm2_vl/modeling.py (hunk 1 above)
Comment (yicycyc) [reply paddleformers/transformers/lfm2_vl/modeling.py:375]: 已在 12e77943 修复：模型 ID、单文件及分片 safetensors 均进入同一 LFM2-VL 转换路径，并增加官方模型 ID 与分片权重回归测试。

File: paddleformers/transformers/lfm2_vl/modeling.py (hunk 2 above)
Comment (yicycyc) [reply paddleformers/transformers/lfm2_vl/modeling.py:317]: 已在 12e77943 修复：pixel_values 存在而 input_ids 缺失时明确拒绝 inputs_embeds 组合并给出可定位错误；新增回归测试。

File: paddleformers/transformers/lfm2_vl/processor.py (hunk 3 above)
Comment (yicycyc) [reply paddleformers/transformers/lfm2_vl/processor.py:71]: 已在 12e77943 修复：文本含 <image> 但 images=None 时立即抛出明确错误，并新增处理器回归测试。

Comment (risemeup1111) [reply paddleformers/transformers/lfm2_vl/processor.py:71]: 已复核当前 head `12e77943`：文本含 `<image>` 且未提供图像时会立即抛出明确错误，新增处理器回归测试覆盖该场景。该 P2 已解决。

File: paddleformers/transformers/lfm2_vl/modeling.py (hunk 1 above)
Comment (risemeup1111) [reply paddleformers/transformers/lfm2_vl/modeling.py:375]: 已复核当前 head `12e77943`：模型 ID、本地单文件和分片 safetensors 现在都经过同一转换路径，新增测试也验证了模型 ID 路由与分片张量读取。该 P1 已解决。

File: paddleformers/transformers/lfm2_vl/modeling.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/lfm2_vl/modeling.py:317]: 已复核当前 head `12e77943`：传入 `pixel_values` 但未提供 `input_ids` 时会在图像注入前明确抛出 `ValueError`，新增回归测试覆盖该路径。该 P1 已解决。

**Review by risemeup1111 — APPROVED** [review]

**Comment by yicycyc** [comment]
本地 A100 已复跑 LFM2-VL 全部 15 项测试，均通过。当前 H20 单卡 CI 失败发生在依赖 Git 拉取阶段（HTTP/2 CANCEL、early EOF），尚未进入模型测试。尝试 rerun-failed-jobs 返回 HTTP 403（需要仓库管理员权限），麻烦维护者协助重跑 run 34306893831。
